Migrate to gem

paoolo opened the pull request.

Nuanda commented on lib/ip_wrangler/nat.rb:
```diff
@@ -0,0 +1,131 @@
+        @iptables.delete_nat_port(nat_port[:public_ip], nat_port[:public_port],
+          nat_port[:private_ip], nat_port[:private_port],
+          nat_port[:protocol])
+        released_port.push({public_ip: nat_port[:public_ip],
```
Aaaaa, give me my inbox back! ;)

pasuder commented:
@dice-cyfronet/atmo-dev-team would like to review it? I will refactor code in the next PRs. The main goal of this PR was migrate to gem structure.

mkasztelnik commented on CHANGELOG:
```diff
@@ -0,0 +1,2 @@
+v 0.0.0
+  - Init
```
I propose more descriptive information here. What is more maybe we can start with 0.1.0 version?

pasuder replied:
OK. I change it.

mkasztelnik commented on bin/ip-wrangler-configure:
```diff
@@ -0,0 +1,60 @@
+    echo "-----"
+    cat config.yml
+
+popd 2>&1 >> /dev/null
```
Heh I see bash fan here :smiling_imp: I know that it is very readable for you with black belt in bash magic but for normal ruby user is can be a little bit scary

pasuder replied:
Bash scripts refactor?

mkasztelnik replied:
Maybe not now but in the future I propose to switch into ruby with more advanced parameter parsing

pasuder replied:
That's true. Currently I extracted function in this scripts.

mkasztelnik commented on bin/ip-wrangler-stop:
```diff
@@ -0,0 +1,10 @@
+export __dir=$( dirname ${__dir} )
+
+pushd ${__dir}/lib/ 2>&1 >> /dev/null
+    thin stop
```
So as far as I understand it only one ipwrangler can be started per user. In the feature app pid file path can be defined in `config.yml` and used while starting/stopping app.

pasuder replied:
One app per installation.

mkasztelnik commented:
While building gem following warnings are reported:

```
λ leeloo ip-wrangler → λ git task_create_gem → gem build ip-wrangler.gemspec 
WARNING:  open-ended dependency on sinatra (>= 0) is not recommended
  if sinatra is semantically versioned, use:
    add_runtime_dependency 'sinatra', '~> 0'
WARNING:  open-ended dependency on thin (>= 0) is not recommended
  if thin is semantically versioned, use:
    add_runtime_dependency 'thin', '~> 0'
WARNING:  open-ended dependency on sequel (>= 0) is not recommended
  if sequel is semantically versioned, use:
    add_runtime_dependency 'sequel', '~> 0'
WARNING:  open-ended dependency on sqlite3 (>= 0) is not recommended
  if sqlite3 is semantically versioned, use:
    add_runtime_dependency 'sqlite3', '~> 0'
WARNING:  open-ended dependency on json (>= 0) is not recommended
  if json is semantically versioned, use:
    add_runtime_dependency 'json', '~> 0'
WARNING:  open-ended dependency on bundler (>= 0, development) is not recommended
  if bundler is semantically versioned, use:
    add_development_dependency 'bundler', '~> 0'
WARNING:  open-ended dependency on rake (>= 0, development) is not recommended
  if rake is semantically versioned, use:
    add_development_dependency 'rake', '~> 0'
WARNING:  See http://guides.rubygems.org/specification-reference/ for help
  Successfully built RubyGem
  Name: ipwrangler
  Version: 0.0.1
  File: ipwrangler-0.0.1.gem
```

It is easily fixable - just set minimum version for used gems in gemspec.

mkasztelnik commented:
I see that generated gem is named: `ipwrangler-0.0.1.gem`. Maybe we should rename github project to `ipwrangler` to be consistent?

mkasztelnik commented:
`Readme.md` update describing how to install gem and user commands in `bin` are missing.

mkasztelnik commented:
Unable to execute `ip-wrangler-configure`:

```
λ leeloo tmp → ip-wrangler-configure 
/home/marek/.rbenv/versions/2.2.0/bin/ip-wrangler-configure:23:in `load': /home/marek/.rbenv/versions/2.2.0/lib/ruby/gems/2.2.0/gems/ipwrangler-0.0.1/bin/ip-wrangler-configure:3: `$(' is not allowed as a global variable name (SyntaxError)
        from /home/marek/.rbenv/versions/2.2.0/bin/ip-wrangler-configure:23:in `<main>'
```

mkasztelnik commented:
For the sake of clarity, maybe all executable in `bin` can be should be renamed from `ip-wrangler-*` into `ipwrangler-*`?

I'm stopping my tests here since executables does not work.

paoolo commented:
I will use name `ip-wrangler`. Less changes.

paoolo commented:
@mkasztelnik updated [`gemspec`](https://github.com/dice-cyfronet/ip-wrangler/blob/task_create_gem/ip-wrangler.gemspec), added ruby-bash wrappers to execute bash scripts, renamed name to `ip-wrangler`, updated [`CHANGELOG`](https://github.com/dice-cyfronet/ip-wrangler/blob/task_create_gem/CHANGELOG), `bins` are able to start.

mkasztelnik commented:
Now `ip-wrangler-configure` works :smile:. Unfortunately it generates `config.yml` in directory where gem is installed:

```
Show /home/marek/.rbenv/versions/2.2.0/lib/ruby/gems/2.2.0/gems/ip-wrangler-0.0.1/lib/config.yml. You may edit this file to add IP which will use to IP mapping.
```

this is bad since gem can be installed in the space not accessible by the user. Try to generate config.yml in local directory and use it while starting `ip-wrangler`

paoolo commented:
OK. I will try to create local `config.yml` and allow to use it.

mkasztelnik commented:
2 additional elements:
- please change cloning repo into normal gem install procedure
- move upstart configuration description into `support/upstart.md` file.

pasuder commented:
Upstart will be not supported. Initd script will be provided.

pasuder commented:
Done. Cleared PR from done proposals.

mkasztelnik commented:
:+1: :clap::clap::clap:

pasuder commented on lib/ip_wrangler/exec.rb:
```diff
@@ -0,0 +1,17 @@
+
+    def execute_command(command)
+      output = `#{command}`
+      puts "Execute: #{command} => output: #{output}, result: #{$?.exitstatus}"
```
You are great!
